hystrix-contrib/hystrix-javanica/hystrix-javanica-core/src/main/java/com/netflix/hystrix/contrib/javanica/command/MetaHolder.java

```diff
@@ -15,96 +15,94 @@
-@Immutable
-public final class MetaHolder {
-
-    private final HystrixCollapser hystrixCollapser;
-    private final HystrixCommand hystrixCommand;
-    private final DefaultProperties defaultProperties;
-
-    private final Method method;
-    private final Method cacheKeyMethod;
-    private final Method ajcMethod;
-    private final Method fallbackMethod;
-    private final Object obj;
-    private final Object proxyObj;
-    private final Object[] args;
-    private final Closure closure;
-    private final String defaultGroupKey;
-    private final String defaultCommandKey;
-    private final String defaultCollapserKey;
-    private final String defaultThreadPoolKey;
-    private final ExecutionType executionType;
-    private final boolean extendedFallback;
-    private final ExecutionType collapserExecutionType;
-    private final ExecutionType fallbackExecutionType;
-    private final boolean fallback;
-    private boolean extendedParentFallback;
-    private final boolean defaultFallback;
-    private final JoinPoint joinPoint;
-    private final boolean observable;
-    private final ObservableExecutionMode observableExecutionMode;
-
-    private static final Function identityFun = new Function<Object, Object>() {
-        @Nullable
-        @Override
-        public Object apply(@Nullable Object input) {
-            return input;
-        }
-    };
-
-    private MetaHolder(Builder builder) {
-        this.hystrixCommand = builder.hystrixCommand;
-        this.method = builder.method;
-        this.cacheKeyMethod = builder.cacheKeyMethod;
-        this.fallbackMethod = builder.fallbackMethod;
-        this.ajcMethod = builder.ajcMethod;
-        this.obj = builder.obj;
-        this.proxyObj = builder.proxyObj;
-        this.args = builder.args;
-        this.closure = builder.closure;
-        this.defaultGroupKey = builder.defaultGroupKey;
-        this.defaultCommandKey = builder.defaultCommandKey;
-        this.defaultThreadPoolKey = builder.defaultThreadPoolKey;
-        this.defaultCollapserKey = builder.defaultCollapserKey;
-        this.defaultProperties = builder.defaultProperties;
-        this.hystrixCollapser = builder.hystrixCollapser;
-        this.executionType = builder.executionType;
-        this.collapserExecutionType = builder.collapserExecutionType;
-        this.fallbackExecutionType = builder.fallbackExecutionType;
-        this.joinPoint = builder.joinPoint;
-        this.extendedFallback = builder.extendedFallback;
-        this.defaultFallback = builder.defaultFallback;
-        this.fallback = builder.fallback;
-        this.extendedParentFallback = builder.extendedParentFallback;
-        this.observable = builder.observable;
-        this.observableExecutionMode = builder.observableExecutionMode;
-    }
-
-    public static Builder builder() {
-        return new Builder();
-    }
+public abstract class MetaHolder<M extends MetaHolder<M,B>,B extends MetaHolder.Builder<M, B>> {
```

Comment:
There are two parts to this question. 

1) Whether MetaHolder should be generic?  
    Need not be. MetaHolder was made generic because of `constructor` and `copy` method. As you mentioned, the same behavior can be accomplished by  passing `MetaHolder.Builder` to constructor and returning `MetaHolder.Builder` from `copy` method. The only down side here is compiler warning about  unchecked conversion.

2) Whether MetaHolder.Builder should be generic?
    It should be generic and the signature should be something like `public abstract static class Builder<M extends MetaHolder, B extends Builder<M, B>>`. I think there is no confusion around `B` as it is required for inheritable builder just like you mentioned in `Foo` example above. My point is that `M` is required for `build` method. `build` method cannot return just MetaHolder. To get a better picture, see how MetaHolder.Builder is used in `AbstractCommandMetaHolderBuilder` which is given below.

```java
public abstract class AbstractCommandMetaHolderBuilder<T extends MetaHolder, V extends MetaHolder.Builder<T,V>> {
    protected final V builder;
    public T build() {
        // Here, If builder returns MetaHolder then we need to cast it to T which is not correct in my opinion.
         return builder.build();
    }
}
```

I hope I have explained it better :) Let me know if I need to explain further.